README.md
# suspend-blocker

The suspend-blocker tool parses the kernel log and will explain the reasons that suspend was woken up or which wakelocks blocked suspend from working. It is useful on Android devices that use wakelocks.

## suspend-blocker command line options

* -b list blocking wakelock names and count
* -h show help
* -H histogram of times between suspend and suspend duration
* -r list causes of resume
* -v verbose information 

##Example Output:

```
suspend-blocker -r -H  < /var/log/kern.log 
stdin:
Resume wakeup causes:
  IRQ 177, nfc_irq                  681  47.82%
  IRQ 162, bcmsdh_sdmmc             678  47.61%
  IRQ 163, gpio_keys                 29   2.04%
  I/O pad: CONTROL_PADCONF_WAK       20   1.40%
  IRQ 39, TWL6030-PIH                13   0.91%
  wakeup I/O pad: CONTROL_WKUP        1   0.07%

Period of time between each successful suspend:
     0.000 -    0.124 seconds        15   2.02%
     0.125 -    0.249 seconds         0   0.00%
     0.250 -    0.499 seconds         0   0.00%
     0.500 -    0.999 seconds         3   0.40%
     1.000 -    1.999 seconds       459  61.86%
     2.000 -    3.999 seconds       111  14.96%
     4.000 -    7.999 seconds        49   6.60%
     8.000 -   15.999 seconds        25   3.37%
    16.000 -   31.999 seconds        13   1.75%
    32.000 -   63.999 seconds         9   1.21%
    64.000 -  127.999 seconds        36   4.85%
   128.000 -  255.999 seconds        11   1.48%
   256.000 -  511.999 seconds         6   0.81%
   512.000 -          seconds         2   0.27%

Duration of successful suspends:
     0.000 -    0.124 seconds        14   1.88%
     0.125 -    0.249 seconds        15   2.02%
     0.250 -    0.499 seconds        29   3.90%
     0.500 -    0.999 seconds        27   3.63%
     1.000 -    1.999 seconds        41   5.52%
     2.000 -    3.999 seconds        56   7.54%
     4.000 -    7.999 seconds        68   9.15%
     8.000 -   15.999 seconds        78  10.50%
    16.000 -   31.999 seconds       132  17.77%
    32.000 -   63.999 seconds       190  25.57%
    64.000 -  127.999 seconds        90  12.11%
   128.000 -  255.999 seconds         3   0.40%

Stats:
  74 suspends aborted (9.06%).
  743 suspends succeeded (90.94%).
  33.200499 seconds average suspend duration (min 0.001000, max 250.768000).
```


debian/changelog
suspend-blocker (0.02.02) xenial; urgency=medium

  * Makefile: bump version
  * Debian: update debhelper to version 9
  * Update Debian standards version
  * Update debian compat to 9
  * Update regression test data
  * Calc stats only on valid successful suspends
  * Add -f option
  * Guard against -ve duration and deltas

 -- Colin King <colin.king@canonical.com>  Mon, 7 Nov 2015 19:05:00 +0000

suspend-blocker (0.02.01) xenial; urgency=medium

  * Makefile: bump version
  * Add syslog.awk to make dist rule
  * free wakeup_sources
  * Update regression tests
  * Add active wakeup sources information
  * Only count wake up messages that are prefixed by [SPM]
  * Add syslog event binning script

 -- Colin King <colin.king@canonical.com>  Thu, 26 Nov 2015 13:37:00 +0000

suspend-blocker (0.02.00) wily; urgency=medium

  * Makefile: bump version
  * Use a more efficient hash function

 -- Colin King <colin.king@canonical.com>  Tue, 14 Jul 2015 13:31:00 +0100

suspend-blocker (0.01.26) vivid; urgency=medium

  * Makefile: bump version
  * Add some scanf format size specifiers
  * Reduce scope of variable valid
  * Remove SIGSEGV trap
  * Default to -O2
  * Initialiase array to stop gcc warning:

 -- Colin King <colin.king@canonical.com>  Wed, 1 Jul 2015 10:18:00 +0100

suspend-blocker (0.01.25) vivid; urgency=medium

  * Makefile: remove -g cc flag, bump version
  * Reduce scope on variable ret
  * Guard against string sscanf buffer overflows
  * Remove signal handling of SIGABRT, SIGBUS, SIGILL
  * Manual: update copyright and date
  * Makefile: add copyright

 -- Colin King <colin.king@canonical.com>  Mon, 18 May 2015 17:17:17 +0100

suspend-blocker (0.01.24) vivid; urgency=medium

  * Makefile: bump version
  * Update tests now that stats are fixed
  * Initialise interval_max, bug found by CoveritScan

 -- Colin King <colin.king@canonical.com>  Fri, 13 Feb 2015 11:46:00 +0000

suspend-blocker (0.01.23) vivid; urgency=medium

  * Makefile: bump version
  * Check sscanf return, ensure we bump counter of wakelock is scanned
  * Fix a couple of spelling mistakes in comments
  * Make a few more functions static
  * Handle gettimeofday errors
  * Improve signal handling
  * Ignore return of fclose

 -- Colin King <colin.king@canonical.com>  Mon, 22 Dec 2014 23:43:00 +0000

suspend-blocker (0.01.22) utopic; urgency=low

  * Makefile: bump version
  * Bucket data in decades (x 10) rather than in powers of two (-d option)
  * Annotate histogram
  * Calculate median and mode correctly
  * Clear whence time stamp after we have dumped out the whence_text field

 -- Colin King <colin.king@canonical.com>  Tue, 19 Aug 2014 12:14:12 +0100

suspend-blocker (0.01.21) utopic; urgency=low

  * Makefile: bump version
  * Fix mixed pm and non-pm timestamps where one set is not valid

 -- Colin King <colin.king@canonical.com>  Mon, 18 Aug 2014 21:04:01 +0100

suspend-blocker (0.01.20) utopic; urgency=low

  * Makefile: bump version
  * Remove list free in histogram

 -- Colin King <colin.king@canonical.com>  Mon, 18 Aug 2014 19:21:23 +0100

suspend-blocker (0.01.19) utopic; urgency=low

  * Makefile: bump version
  * Fix leak on multiple suspend_fail_cause strdups
  * Free lists
  * Update tests
  * Enable for new phone support with older 3.4.x kernel
  * Add more device support

 -- Colin King <colin.king@canonical.com>  Mon, 18 Aug 2014 17:05:01 +0100

suspend-blocker (0.01.18) utopic; urgency=low

  * Makefile: bump version
  * Fix incorrect strcmp, compare on different strings

 -- Colin King <colin.king@canonical.com>  Tue, 29 Apr 2014 22:08:12 +0100

suspend-blocker (0.01.17) trusty; urgency=low

  * Makefile: bump version
  * fix json output when taking data from stdin

 -- Colin King <colin.king@canonical.com>  Sat, 5 Apr 2014 10:37:28 +0100

suspend-blocker (0.01.16) trusty; urgency=low

  * Makefile: bump version
  * Allow suspend-blocker to be interrupted and dump stats immediately

 -- Colin King <colin.king@canonical.com>  Thu, 3 Mar 2014 14:28:41 +0100

suspend-blocker (0.01.15) trusty; urgency=low

  * Makefile: bump version
  * Fix out of bounds array indexing error found by Coverity Scan
  * Fix json errors found by Coverity Scan
  * Reduce variable scopes
  * Makefile: Fix make dist, use plain old tar to omit debian contents
  * Add COPYING file

 -- Colin King <colin.king@canonical.com>  Thu, 20 Feb 2014 15:57:22 +0000

suspend-blocker (0.01.14) trusty; urgency=low

  * Makefile: bump version again
  * Add example python script to parse json output and pass/fail the results
  * ensure each suspend kernel log parse is an individual json results object
  * Add example script to show how to use the tool

 -- Colin King <colin.king@canonical.com>  Fri, 7 Feb 2014 17:28:01 +0000

suspend-blocker (0.01.13) trusty; urgency=low

  * Makefile: bump version
  * Update copyright
  * Cater for suspend wakeups for Nexus 10, fix counter bug
  * Add in support for newer /sys wakelock interface
  * Update debian standards version

 -- Colin King <colin.king@canonical.com>  Fri, 7 Feb 2014 14:38:00 +0000

suspend-blocker (0.01.12) trusty; urgency=low

  * Makefile: bump version
  * Calculate stats based on real time elapsed rather than requested time
  * Check for weird -ve time issues
  * Add json output to -o option
  * Add -o option to help
  * Update options and help with new -w option
  * Add -W option

 -- Colin King <colin.king@canonical.com>  Mon, 3 Feb 2014 16:42:00 +0000

suspend-blocker (0.01.11) saucy; urgency=low

  * Makefile: bump version
  * Update tests
  * Add new options -q and -o and more stats reporting
  * Add -o filename option to dump results in JSON format to a file
  * Add -q quiet mode
  * Add more stats reporting (mean, median, mode, etc)

 -- Colin King <colin.king@canonical.com>  Mon, 4 Nov 2013 18:02:11 +0000

suspend-blocker (0.01.10) saucy; urgency=low

  * Makefile: update version, remove -g flag
  * Ensure min/max are sane with null input


test.sh
#!/bin/bash

for I in *.klog
do
	./suspend-blocker -v -b -r < $I > /tmp/$I.output
	diff $I.output /tmp/$I.output
	if [ $? -eq 0 ]; then
		echo "$I: PASSED"
	else
		echo "$I: FAILED"
	fi
done


scripts/example.sh
#!/bin/bash
#
# Copyright (C) 2014 Canonical, Ltd.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#

DURATION=60

if [ $EUID -ne 0 ];
then
	echo must be root to run this script
	exit 0
fi

#
# Clean kernel log, restart rsyslogd
#
rm /var/log/kern.log
kill -HUP $(cat /var/run/rsyslogd.pid)

#
#  Collect wakelock stats
#
./suspend-blocker -w $DURATION -o wakelock-stats.json
#
#  And parse kernel log for suspend reasons
#
./suspend-blocker /var/log/kern.log -r -o suspend-stats.json




scripts/syslog.awk
#!/usr/bin/awk -f
#
# Copyright (C) 2015 Canonical, Ltd.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#

BEGIN {
	m_min = 1440
	m_max = 0
}

{
	t=$3
	h=substr(t, 0, 2)
	m=substr(t, 4, 2)
	mins=(h * 60) + m

	if (mins > m_max)
		m_max = mins
	if (mins < m_min)
		m_min = mins
	x=index($0, "wake up by");
	if (x > 0) {
		s=substr($0, x + 11)
		y=index(s, ",")
		if (y > 1) {
			wakeup = substr(s, 0, y - 1)
			wakeups[wakeup]++
			event[mins][wakeup]++
			times[mins]++
			time[mins] = substr(t, 0, 5)
			#print "-->", mins, wakeup, event[mins][wakeup]
		}
	}
}

END {	
	printf "Time HH:MM\t"
	for (w in wakeups)
		printf "%s\t", w
	printf "\n"

	for (m = m_min; m < m_max; m++) {
		if (times[m] > 0) {
			printf "%s\t", time[m]
			for (w in wakeups) {
				if (event[m][w] > 0)
					printf "%d\t", event[m][w]
				else
					printf "0\t"
			}
			printf "\n"
		}
	}
}
	


scripts/parse-json.py
#!/usr/bin/python
#
# Copyright (C) 2014 Canonical, Ltd.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, 
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
#
import json
import sys
import os

debug = False

def parse_wakelock_data(data):
    if debug:
        print "Parsing results from wakelocks.."
        print "Test duration (seconds): %f" % data['duration-seconds']

    if data['duration-seconds'] < 60:
        print "FAILED: %s test duration was less than 60 seconds: %f" % data['duration-seconds']
    for wl in data['wakelocks']:
        if debug:
            print "Wakelock: %s" % wl['wakelock']
            print "  Active count (per second): %f" % wl['active_count_per_second']
            print "  Count (per second)       : %f" % wl['count_per_second']
            print "  Expire count (per second): %f" % wl['expire_count_per_second']
            print "  Wakeup count (per second): %f" % wl['wakeup_count_per_second']
            print "  Total time (%%)           : %f" % wl['total_time_percent']
            print "  Sleep time (%%)           : %f" % wl['sleep_time_percent']
            print "  Prevent time (%%)         : %f" % wl['prevent_time_percent']
            print " "

        if wl['total_time_percent'] > 5.0:
            print "FAILED: %s total time too large: %f %%" % (wl['wakelock'], wl['total_time_percent'])

def parse_wakelock_klog(data):
    if debug:
        print "Parsing results from klog analysis.."
    #
    #  We can have one or more kernel logs to parse
    #
    for kernlog in data:
        if debug:
           print "Kernel log: %s" % kernlog['kernel-log']
           print "Suspends attempted                      : %d" % kernlog['suspends-attempted']
           print "Suspends aborted                        : %d" % kernlog['suspends-aborted']
           print "Suspends succeeded                      : %d" % kernlog['suspends-succeeded']
           print "Suspends aborted (%%)                    : %f" % kernlog['suspends-aborted-percent']
           print "Suspends succeeded (%%)                  : %f" % kernlog['suspends-succeeded-percent']
           print "Suspends total time (%%)                 : %f" % kernlog['suspends-total-time-percent']
           print "Suspend maximum duration (seconds)      : %f" % kernlog['suspend-maximum-duration-seconds']
           print "Awake maximum duration (seconds)        : %f" % kernlog['awake-maximum-duration-seconds']

        if kernlog['suspends-aborted-percent'] > 25:
            print "FAILED: %s: aborted suspends too large: %f" % (kernlog['kernel-log'], kernlog['suspends-aborted-percent'])
        if kernlog['suspends-succeeded-percent'] < 75:
            print "FAILED: %s: succeeded suspends too small: %f" % (kernlog['kernel-log'], kernlog['suspends-succeeded-percent'])
        if kernlog['suspends-attempted'] < 1:
            print "FAILED: %s: did not attempt to suspend, this is unexpected." % kernlog['kernel-log']
        if kernlog['suspends-succeeded'] < 1:
            print "FAILED: %s: did not succeed any suspends, most probably blocked by a wakelock." % kernlog['kernel-log']
        if kernlog['suspend-maximum-duration-seconds'] < 30:
            print "FAILED: %s: maximum duration of suspend was %f seconds, way too short." % (kernlog['kernel-log'], kernlog['suspend-maximum-duration-seconds'])
        if kernlog['awake-maximum-duration-seconds'] > 10:
            print "FAILED: %s: maximum duration of awake time was %f seconds, way too long." % (kernlog['kernel-log'], kernlog['awake-maximum-duration-seconds'])
    

def main(file):
    with open(file) as f:
        data = json.load(f)

    if data.has_key('wakelock-data'):
        parse_wakelock_data(data['wakelock-data'])
    if data.has_key('wakelock-stats-from-klog'):
        parse_wakelock_klog(data['wakelock-stats-from-klog'])


if __name__ == '__main__':
    if len(sys.argv) != 2:
        sys.stderr.write("Usage: %s jsonfile\n" % sys.argv[0])
        exit(1)

    main(sys.argv[1])
